runtime/src/lib.rs

```diff
@@ -19,353 +19,476 @@
+/// The type for looking up accounts. We don't expect more than 4 billion of them, but you
+/// never know...
+pub type AccountIndex = u32;
 
-/// A hash of some data used by the chain.
-pub type Hash = primitives::H256;
+/// Balance of an account.
+pub type Balance = u128;
 
-/// Index of a block number in the chain.
-pub type BlockNumber = u64;
+/// Index of a transaction in the chain.
+pub type Index = u64;
+
+/// A hash of some data used by the chain.
+pub type Hash = sp_core::H256;
 
-/// Index of an account's extrinsic in the chain.
-pub type Nonce = u64;
+/// Digest item type.
+pub type DigestItem = generic::DigestItem<Hash>;
 
 /// Opaque types. These are used by the CLI to instantiate machinery that don't need to know
 /// the specifics of the runtime. They can then be made to be agnostic over specific formats
 /// of data like extrinsics, allowing for them to continue syncing the network through upgrades
 /// to even the core data structures.
 pub mod opaque {
-	use super::*;
-
-	/// Opaque, encoded, unchecked extrinsic.
-	#[derive(PartialEq, Eq, Clone, Default, Encode, Decode)]
-	#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
-	pub struct UncheckedExtrinsic(#[cfg_attr(feature = "std", serde(with = "bytes"))] pub Vec<u8>);
-	#[cfg(feature = "std")]
-	impl std::fmt::Debug for UncheckedExtrinsic {
-		fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
-			write!(fmt, "{}", primitives::hexdisplay::HexDisplay::from(&self.0))
-		}
-	}
-	impl traits::Extrinsic for UncheckedExtrinsic {
-		fn is_signed(&self) -> Option<bool> {
-			None
-		}
-	}
-	/// Opaque block header type.
-	pub type Header = generic::Header<
-		BlockNumber,
-		BlakeTwo256,
-		generic::DigestItem<Hash, AuthorityId, AuthoritySignature>,
-	>;
-	/// Opaque block type.
-	pub type Block = generic::Block<Header, UncheckedExtrinsic>;
-	/// Opaque block identifier type.
-	pub type BlockId = generic::BlockId<Block>;
-	/// Opaque session key type.
-	pub type SessionKey = AuthorityId;
+    use super::*;
+
+    pub use sp_runtime::OpaqueExtrinsic as UncheckedExtrinsic;
+
+    /// Opaque block header type.
+    pub type Header = generic::Header<BlockNumber, BlakeTwo256>;
+    /// Opaque block type.
+    pub type Block = generic::Block<Header, UncheckedExtrinsic>;
+    /// Opaque block identifier type.
+    pub type BlockId = generic::BlockId<Block>;
+
+    impl_opaque_keys! {
+        pub struct SessionKeys {
+            pub aura: Aura,
+            pub grandpa: Grandpa,
+        }
+    }
 }
 
 /// This runtime version.
 pub const VERSION: RuntimeVersion = RuntimeVersion {
-	spec_name: create_runtime_str!("mashnet-node"),
-	impl_name: create_runtime_str!("mashnet-node"),
-	authoring_version: 3,
-	spec_version: 4,
-	impl_version: 4,
-	apis: RUNTIME_API_VERSIONS,
+    spec_name: create_runtime_str!("mashnet-node"),
+    impl_name: create_runtime_str!("mashnet-node"),
+    authoring_version: 4,
+    spec_version: 4,
+    impl_version: 5,
```

Comment:
Can this start at 0/1 again?

Reply:
Me
> Short question about the incrementing of the runtime version. What would be the recommended best practice when we switch from Substrate 1.x to 2.0?
1. Increment all version numbers
2. "Reset" to 1 if current versions > 1

Jam from Parity

> It only really matters if you have external tooling pointing to your chains and you have some kind of continuity of the infrastructure between versions. I generally don't think it's a good idea to go backwards in version number unless you are changing the name or something.

Reply:
Not sure, if he did understand the questions correctly. We only want to put `impl_version` to 1, while increasing `authoring_version` and `spec_version`.
Like going from `0.2.8` to `0.3.0`.
But we actually do not have the restrictions, he is talking about, so I guess it doesn't matter. Would still be interesting to know, how it is supposed to be done.

Reply:
Gav said something about this: https://github.com/paritytech/substrate/issues/241

Reply:
Also only really matters, when updating the runtime on-chain.